File: ruby/fuzz/targets/verify_receipt.rb
# frozen_string_literal: true

# The whole legacy-receipt path on DER bytes: CMS walk, payload parse, chain
# build, signature check.
#
# Three invariants, the ones the Go port's FuzzVerifyReceipt states:
#
#   * nothing escapes uncaught;
#   * a failure is a VerificationError, never a NoMethodError, a TypeError or
#     a SystemStackError;
#   * an accepted receipt was accepted *because of* the anchors, proven by
#     re-running it against an unrelated anchor set and requiring failure.
#
# Without the third a fuzzer can find crashes but never "accepts what it
# should not". The trusted set is the pinned Apple roots plus the generated
# fixture root, so both the shared fixture receipts and the two public Apple
# receipts get past the chain check and the fuzzer can explore what lies
# beyond it; the unrelated set is the fixture *JWS* root.

require "ruzzy"
require_relative "../support"

APRV = FuzzSupport::APRV

TRUSTED = (APRV.apple_receipt_roots +
           [FuzzSupport.fixture_certificate("generated/receipt-root.der")]).freeze
UNRELATED = [FuzzSupport.fixture_certificate("generated/jws-root.der")].freeze

TEST_ONE_INPUT = lambda do |data|
  outcome, receipt = FuzzSupport.call("verify_receipt_core", APRV::VerificationError) do
    APRV.verify_receipt_core(data, trusted_roots: TRUSTED)
  end
  next nil unless outcome == :accepted

  FuzzSupport.violated("verify_receipt_core returned nil for an accepted receipt") if receipt.nil?

  again, = FuzzSupport.call("verify_receipt_core (unrelated anchors)", APRV::VerificationError) do
    APRV.verify_receipt_core(data, trusted_roots: UNRELATED)
  end
  if again == :accepted
    FuzzSupport.violated("this input verifies against an unrelated anchor set too, " \
                         "so the anchors are not being enforced")
  end
  nil
end

Ruzzy.fuzz(TEST_ONE_INPUT)

File: php/fuzz/targets/verify-receipt.php
<?php

declare(strict_types=1);

namespace EminDeniz99\ApplePurchaseReceiptVerifier\Fuzz;

use EminDeniz99\ApplePurchaseReceiptVerifier\AppleRootCerts;
use EminDeniz99\ApplePurchaseReceiptVerifier\Receipt\ReceiptVerifier;
use EminDeniz99\ApplePurchaseReceiptVerifier\VerificationException;

/**
 * The whole legacy-receipt path on DER bytes: CMS walk, payload parse, chain
 * build, signature check.
 *
 * The invariants are the three the Go port's `FuzzVerifyReceipt` states:
 * nothing escapes but `VerificationException` (an `InvalidArgumentException`
 * is unreachable with a non-empty anchor set, so it is a bug here, and so is
 * a `TypeError` or a warning from a parser reached with an unexpected node
 * shape); and an accepted receipt is accepted BECAUSE of the anchors, proven
 * by re-running it against an unrelated anchor set and requiring failure.
 * Without that third one a fuzzer can find crashes but never "accepts what it
 * should not".
 *
 * The anchor set is the pinned Apple roots plus the generated fixture receipt
 * root, so both the shared fixture receipts and the two public Apple receipts
 * get past the chain check and the fuzzer can explore what lies beyond it.
 * The unrelated set is the fixture *JWS* root.
 */

/** @var \PhpFuzzer\Config $config */
require __DIR__ . '/../bootstrap.php';

$trusted = FuzzFixtures::withReceiptRoot(AppleRootCerts::receiptRoots());
$unrelated = FuzzFixtures::jwsRootOnly();

$config->setMaxLen(16384);
$config->setAllowedExceptions([VerificationException::class]);

$config->setTarget(static function (string $input) use ($trusted, $unrelated): void {
    ReceiptVerifier::verifyReceiptCore($input, $trusted);

    // Reached only when the receipt verified. An Error, not an exception:
    // the fuzzer treats Errors as findings and Exceptions as verdicts.
    try {
        ReceiptVerifier::verifyReceiptCore($input, $unrelated);
    } catch (VerificationException) {
        return;
    }

    throw new \Error(
        'this input verifies against an unrelated anchor set too, '
        . 'so the anchors are not being enforced',
    );
});
